### DISCUSSION TITLE
macOS and Mobile UX Roadmap

### DISCUSSION THREAD

**Discussion by unphased** [Comfy-Org/ComfyUI#2764] [General]
I'm writing this to continue the thread and provide an overview and up-to-date checklist for a number of related things that are of interest to my workflows. For a bit of background I run ComfyUI in Linux and I'm currently getting set up for StableSwarmUI to be able to leverage all three of my Ampere RTX GPUs. For the past several months my experience has been that it is not ergonomic to manually toggle between 3 comfy instance tabs to keep all the GPUs busy.

 I do most of my experiments in ComfyUI from my macbook, and lately I'm really excited about exploring streamlined live editing using my iPad leveraging comfyui workflows by scribbling controlnet compositions with the Apple Pencil! It goes without saying that once the UX gets there for mobile in this frontend, the iPad should naturally transform into the platform of choice. 

You can follow my initial written ideas i had put [here](https://github.com/comfyanonymous/ComfyUI/pull/109#issuecomment-1936905814): 

### Basic platform UX

- [x] https://github.com/danphilibin/TinkerBot-tech-for-ComfyUI-Touchpad/tree/main <-- implements "custom node" to give macOS touchpad two finger X/Y pan and pinch to zoom. No longer a pain point, so I'm checking off the box, but I think this should be incorporated into comfyui out of the box and not existing as a separate plugin.
- [x] Safari rendering freeze, I submitted the fix for this today in #2763 -- this fixes the performance issue in rendering the search box to be usable from Apple mobile devices as well as macOS Safari
- [x] Basic Apple mobile device multitouch capability has already been merged in #2616. This enables two finger pinch to zoom, a bit of critical mobile functionality
- [ ] I aim to contribute in the next days: **Multitouch two finger combination pinch to zoom and pan**, a further improvement that brings more fluid navigation and perfect canvas-to-finger positional mapping in the mobile device navigation UX. There is no reason for comfyui not to be every bit as fluid as any maps app.
   - [ ] Add momentum to pan with Apple Maps behavior as baseline. Make a poll here to see if anyone likes it. Possibly hide it as opt-in, behind a config.
   - [ ] Consider adding doubletap-drag up and down as one-handed-phone-use-case accessibility gestures for zooming out and in. (Did you know? The direction of this not-well-known gesture is reversed between Apple Maps and Google Maps, but both have it) If I implement this feature I would add further client detection and enable it only on phone class devices.
- [ ] There is a small tweak I would make to the behavior seen in `TinkerBot-tech-for-ComfyUI-Touchpad` around some of the mathematical equations (to be specific, it's overly sensitive when you are zoomed out and overly insensitive when zooming in). I will also move it into comfyui to improve the status quo of macbook users being handicapped behind the installation of this separate plugin that nobody knows about (yet?).
- [ ] iPhone: viewport allowed to scale right now. Needs to bring additional strategies into the fold to prevent this behavior because user-scalable=no in meta tag is not working. 
- [ ] iPhone: tapping in many places such as Generate button and stuff like the Settings menu checkboxes to respond twice making it nearly impossible to actually toggle the checkboxes for example. Virtual events should be banished.
- [ ] brainstorm what else might make sense to tweak in terms of UX when we detect an apple pencil/stylus is present. Examples of ideas to get us started:
   - [ ] Suppose we go with a paradigm in which two finger drag will always pan, one finger drag will work like mouse drag (either pan canvas or drag nodes), then stylus could be released from the job of moving nodes, and instead we can make the hitboxes for the noodle sockets on nodes as large as possible for the stylus to hit, then that will maximize the speed that one could use a stylus to rewire nodes, as well as making a clear distinction to use fingers to move nodes and stylus to rewire them. 
   - [ ] Another idea (probably this would want a new button and state toggle in the main menu?) Use stylus to perform selection in the canvas. Starting point is just letting it do the same thing Ctrl+drag does now, so you could then multiselect without a keyboard.
   - [ ] ..., etc. With all these the presence of a stylus gives a way to streamline and make interaction more powerful/flexible, but should not lock a user with only finger input out. For example for selection, when you hit the button without a stylus, then behavior would change to just taking the next drag as a selection.

Goes without saying and as a note to self, test all these changes for nonbreakage on established platforms. The ones I'm well set up to test are:
- chrome, firefox on windows
- chrome, firefox, safari on macOS
- firefox on Ubuntu

Implicit from the above but I'll list:
- mobile Safari on iPhone
- mobile Safari on iPad with & without use of Apple Pencil

----- 

### Further ambitions

I aim to develop a simple painting node with native browser stylus API (and which also will work with just your finger of course) to add Apple Pencil support into a custom node for ComfyUI. This will be most likely done outside of this repo but it's possible that certain capabilities may require special changes for integration. This will be a moderate effort but well worth it for direct control image drawing workflow within ComfyUI. I have now experimented with some Krita plugins and it's nice but the separation from doing it all in comfyui is a non-starter for me. 

- [ ] Prefer to implement paint canvas by leveraging comfyui workflow zoom seamlessly
- [ ] (stretch goal) Add ability to apply a temporary rotational transform to the paint canvas to assist in drawing
- [ ] Provide ability to overlay and properly align generation output to the paint canvas. Maybe this is gonna look like some full blown layers system, maybe not. But I'm talking about building whatever it is that we would need to be able to have slick live painting workflows. 
   - [ ] I still don't know what ideal flow is. I experimented in Krita a bit with opacity on the generation so I can see the scribble underneath. But it's far from ideal. Probably having a big button to toggle may be a superior workflow. Or I'm just kidding myself and the best is gonna just be side by side, which is trivial.
   - [ ] look into whether browser api allows access to Apple Pencil Double Tap gesture and use for the above if possible.
- [ ] Support scribble image painting
- [ ] Support for painting and editing masks (the scope of this will expand further...)
   - [ ] lasso
   - [ ] SAM and other segmentation
   - [ ] paintbrush on alpha mask
- [ ] More of the usual suspects: color picker, eraser, button to clear, paint bucket flood fill, brush config/pressure sensitivity, etc. 
- [ ] Support pose editing (brainstorm: I think if using stylus, can enable edges to be draggable as well)
- [ ] etc. (wishlist: 3d models/editor etc.)

**Comment by fcogomez** [comment]
Do you mind fixing the load image node? It doesn't work on iPad (iOS). Thanks

**Reply by unphased** [reply]
Thanks for the bug report! I will look into it when I get a chance.
